circuit/src/androidMain/kotlin/com/slack/circuit/Navigator.android.kt

```diff
@@ -18,38 +18,43 @@ package com.slack.circuit
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.remember
 import com.slack.circuit.backstack.SaveableBackStack
+import com.slack.circuit.backstack.isEmpty
 
 /**
  * Returns a new [Navigator] for navigating within [CircuitContents][CircuitContent].
  *
  * @see NavigableCircuitContent
  *
  * @param backstack The backing [SaveableBackStack] to navigate.
- * @param onRootPop The callback to handle root [Navigator.pop] calls.
+ * @param onRootPop A callback to handle when the [backstack] is [popped][Navigator.pop] to empty.
  */
 @Composable
 public fun rememberCircuitNavigator(
   backstack: SaveableBackStack,
-  onRootPop: (() -> Unit)?
+  onRootPop: (() -> Unit)
 ): Navigator {
   return remember { NavigatorImpl(backstack, onRootPop) }
 }
 
-private class NavigatorImpl(
+internal class NavigatorImpl(
   private val backstack: SaveableBackStack,
-  private val onRootPop: (() -> Unit)?,
+  private val onRootPop: (() -> Unit),
 ) : Navigator {
 
+  init {
+    check(!backstack.isEmpty) { "Backstack size must not be empty." }
+  }
+
   override fun goTo(screen: Screen) {
     backstack.push(screen)
   }
 
   override fun pop(): Screen? {
-    backstack.pop()?.screen?.let {
-      return it
+    val screen = backstack.pop()?.screen
+    if (backstack.size == 0) {
+      onRootPop()
     }
```

Comment:
Observation: we've made efforts to prevent the user from creating an empty back stack, yet allow them to pop their way to an empty stack. We assume they will remove the activity/fragment hosting the current composables using `onBackPressedDispatcher::onBackPressed` or similar, yet don't really do anything to enforce it. This feels a bit weird to me. If we're going to allow an empty back stack via popping, why not allow it on init?

Reply:
I'm not sure I follow, how are we allowing an empty back stack on pop? This code specifically moves to onRootPop when it hits empty

Reply:
Looking at the code in `MainActivity` I noted that we initialize `NavigatorImpl` with `onRootPop` set to `onBackPressedDispatcher::onBackPressed`. At the time, my assumption was that we were expecting `onRootPop` to perform a system back but we were doing nothing to enforce this, meaning the caller could instead pass an empty lambda and end up in a weird state with an empty back stack. I just tried this and see that I had the wrong impression.

That being said, it doesn't seem that `onRootPop` is _ever_ called (unless I'm missing something). The call to `BackHandler` in `NavigableCircuitContent` monitors the state of the back stack and only calls `Navigator.pop` when the backstack is NOT at the root element. This means the conditional in `pop`:
```
    if (backstack.size == 0) {
      onRootPop()
    }
```
will never be true, meaning `onRootPop` won't ever be run.

Reply:
It does this check immediately after popping

```kotlin
    val screen = backstack.pop()?.screen
    if (backstack.size == 0) {
      onRootPop()
    }
```

Breaking this down with comments for clarity, it looks like this when the root pop callback would be invoked

```kotlin
    // backstack.size is 1
    val screen = backstack.pop()?.screen
    // backstack.size is now 0 after popping
    if (backstack.size == 0) {
      onRootPop()
    }
```

So if you are popping the last item off, it will call onRootPop(). How would that never be run?

Reply:
Finally got a good solution for this that's actually significantly simpler. I've removed root pop handling entirely, and instead the user should just add their own `BackHandler` _before_ creating a navigator. This avoids the reentrant behavior, makes `NavigatorImpl` simpler, and works a bit better with the direction the system UI is trying to go with back handling in general.